agenta-backend/agenta_backend/middlewares/api_key_manager.py

```diff
@@ -0,0 +1,229 @@
+
+                # Check rate limiting in Redis
+                if self.redis_client is None:
+                    await self.init_redis()
+
+                cache_key = f"api_request_count:{key}"
+                api_requests_within_minute = await self.redis_client.get(cache_key)
+
+                if api_requests_within_minute is None:
+                    # Initialize the count in Redis with an initial value of 1 and a one-minute TTL
+                    await self.redis_client.setex(cache_key, 60, 1)
+                else:
+
+                    # Check if requests made within the last minute exceed the rate limit
+                    count_within_minute = int(api_requests_within_minute)
+                    if count_within_minute > api_key.rate_limit:
+                        raise ValueError("API key rate limit exceeded")
+
+            # Update the last usage timestamp for rate limiting
+            api_key.updated_at = datetime.utcnow()
+            engine.save(api_key)
+
+            return True
+        except Exception as e:
+            logger.error(f"Error using API key: {e}")
+            raise e
+
+
+async def authenticate_middleware(request: Request, call_next):
+    """
+    Middleware function responsible for authenticating API requests.
+
+    Args:
+        request (Request): The incoming HTTP request object.
+        call_next: The function to call to proceed to the next middleware or route handler.
+
+    Raises:
+        HTTPException: If the request does not contain a valid API key in the Authorization header.
+
+    Returns:
+        The response from the next middleware or route handler.
+    """
+
+    authorization_key = request.headers.get("Authorization")
+    if not authorization_key:
+        raise HTTPException(status_code=401, detail="API Key is missing")
+
+    # Validate the API key using the APIKeyManager
+    try:
+        APIKeyManager.use_api_key(authorization_key)
+    except ValueError as e:
+        raise HTTPException(status_code=403, detail=str(e))
+    except Exception:
+        raise HTTPException(status_code=403, detail="Couldn't validate API Key")
+
+    # Decrypt the API key using the decrypt_api_key function
+    decrypted_user_id = decrypt_api_key(authorization_key)
+    if not decrypted_user_id:
+        raise HTTPException(status_code=403, detail="Invalid API Key")
+
+    # Store the decrypted API key in the request state for later use
+    request.state.user_id = decrypted_user_id
+
+    response = await call_next(request)
+    return response
+
+
+def decrypt_api_key(authorization_key: str) -> str:
+    """
+    Decrypts the authorization key and returns the decrypted user ID.
+
+    Args:
+        authorization_key (str): The encrypted authorization key that needs to be decrypted.
+
+    Returns:
+        str: The decrypted user ID extracted from the authorization key.
+    """
+    try:    
+        decrypted_user_id = fernet.decrypt(authorization_key.encode()).decode()
```

Comment:
Nope we don't have a function to decrypt the api key. we don't need it

Reply:
please read this https://www.freecodecamp.org/news/best-practices-for-building-api-keys-97c26eabfea9/

Reply:
Please read this too: https://blog.mergify.com/api-keys-best-practice/
Obviously it's too much here. But please see how to create the api keys